plugins/history/history.7
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD-PLUGIN-HISTORY 7
.Os
.\" NAME
.Sh NAME
.Nm history
.Nd irccd history plugin
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
plugin is used to check when someone has been seen for the last time on a
channel. For that purpose, irccd needs to be on that channel forever if
possible.
.\" INSTALLATION
.Sh INSTALLATION
The plugin
.Nm
is distributed with irccd. To enable it add it into your
.Xr irccd.conf 5
configuration file.
.Pp
.Bd -literal -offset indent
plugin "history"
.Ed
.\" USAGE
.Sh USAGE
The plugin
.Nm history
reacts to the special command. It understands
.Ar seen
and
.Ar said
sub commands.
.Pp
.Bl -bullet -compact
.It
The sub command
.Ar seen
tells you when the user has been seen for the last time.
.It
The sub command
.Ar said
tells you what the user has said for the last time.
.El
.Pp
Example
.Bd -literal -offset Ds
markand: !history seen jean
irccd: markand, the last time I've seen jean was on 18/01/1989 at 20:55
markand: !history said jean
irccd: markand, the last thing that jean said was: hello world
.Ed
.\" CONFIGURATION
.Sh CONFIGURATION
The plugin
.Nm
supports the following configuration properties:
.Bl -tag -width "file (string)"
.It Va file No (string)
path to the JSON file for saving information (Optional. default to cache
directory).
.El
.\" TEMPLATES
.Sh TEMPLATES
The plugin
.Nm
supports the following template properties:
.Bl -tag -width unknown
.It Va error
Template when an internal error occured. Keywords:
.Em target , (date) .
.It Va silent
Template when someone has been seen but who has never wrote any message.
Keywords:
.Em target , (date) .
.It Va seen
Template for showing last time someone was seen. Keywords:
.Em target , (date) .
.It Va said
Template for showing the last message. Keywords:
.Em message , target , (date) .
.It Va unknown
Template when the user has never been seen. Keywords:
.Em target .
.It Va usage
Template to show the plugin usage.
.El
.Pp
All templates support the following standard keywords:
.Em channel , command , nickname , origin , plugin , server .
.\" EXAMPLES
.Sh EXAMPLES
Example of configuration file:
.Bd -literal
plugin "history" {
  templates {
    "seen" "#{target} has been seen on #{channel} the last time on: %d/%m/%Y %H:%M";
  }
}
.Ed
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd 1 ,
.Xr irccd-templates 7


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


plugins/history/history.js
/*
 * history.js -- track nickname's history
 *
 * Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// Plugin information.
info = {
	author: "David Demelier <markand@malikania.fr>",
	license: "ISC",
	summary: "track nickname's history",
	version: "@irccd_VERSION@"
};

// Modules.
var Directory = Irccd.Directory;
var File = Irccd.File;
var Logger = Irccd.Logger;
var Plugin = Irccd.Plugin;
var Server = Irccd.Server;
var Util = Irccd.Util;

Plugin.templates = {
	"error":        "#{nickname}, I'm sorry, something went wrong.",
	"seen":         "#{nickname}, I've seen #{target} for the last time the %d-%m-%Y %H:%M",
	"said":         "#{nickname}, #{target} said on %d-%m-%Y %H:%M: #{message}",
	"silent":       "#{nickname}, #{target} hasn't said anything yet.",
	"unknown":      "#{nickname}, I've never seen #{target}.",
	"usage":        "#{nickname}, usage: #{plugin} seen | said <target>."
};

function isSelf(server, origin)
{
	return server.info().nickname === Util.splituser(origin);
}

function command(server)
{
	return server.info().prefix + "history";
}

function path(server, channel)
{
	var p;

	if (Plugin.config["file"] !== undefined) {
		p = Util.format(Plugin.config["file"], {
			"server":       server.toString(),
			"channel":      channel
		});
	} else
		p = Plugin.paths.cache + "/db.json";

	return p;
}

function read(server, channel, nickname)
{
	var p = path(server, channel);
	var db = {};

	if (File.exists(p)) {
		var file = new File(path(server, channel), "r");
		var str = file.read();

		db = JSON.parse(str);
	}

	// Complete if needed.
	if (!db[server])
		db[server] = {};
	if (!db[server][channel])
		db[server][channel] = {};
	if (!db[server][channel][nickname])
		db[server][channel][nickname] = {};

	return db;
}

function write(server, channel, nickname, message)
{
	var db = read(server, channel, nickname);
	var entry = db[server][channel][nickname];
	var p = path(server, channel);

	if (!File.exists(File.dirname(p))) {
		Logger.debug("creating directory " + File.dirname(p));
		Directory.mkdir(File.dirname(p));
	}

	var file = new File(path(server, channel), "wt");

	entry.timestamp = Date.now();

	if (message) {
		entry.message = message;
		entry.messageTimestamp = entry.timestamp;
	}

	file.write(JSON.stringify(db));
}

function find(server, channel, target)
{
	var db = read(server, channel, target);
	var it = db[server][channel][target];

	if (it.timestamp)
		return it;
}

function onCommand(server, origin, channel, message)
{
	channel = channel.toLowerCase();

	var args = message.trim().split(" ");
	var kw = {
		channel: channel,
		command: command(server),
		nickname: Util.splituser(origin),
		origin: origin,
		plugin: Plugin.info().name,
		server: server.toString()
	};

	if (args.length !== 2 || args[0].length === 0 || args[1].length === 0) {
		server.message(channel, Util.format(Plugin.templates.usage, kw));
		return;
	}

	if (args[0] !== "seen" && args[0] !== "said") {
		server.message(channel, Util.format(Plugin.templates.usage, kw));
		return;
	}

	args[1] = args[1].toLowerCase();

	if (isSelf(server, args[1]))
		return;

	try {
		var info = find(server, channel, args[1]);

		kw.target = args[1];

		if (!info) {
			server.message(channel, Util.format(Plugin.templates["unknown"], kw));
			return;
		}

		kw.date = info.timestamp;

		if (args[0] === "seen")
			template = Plugin.templates["seen"];
		else {
			/*
			 * If there is a message, we're more interested in the
			 * message timestamp rather than when he was seen
			 * because someone may be on a channel for a long time
			 * (seen) but hasn't talk for long.
			 */
			if (info.message) {
				kw.message = info.message;
				kw.date = info.messageTimestamp;
				template = Plugin.templates["said"];
			} else
				template = Plugin.templates["silent"];
		}

		server.message(channel, Util.format(template, kw));
	} catch (e) {
		server.message(channel, Util.format(Plugin.templates["error"], kw));
	}
}

function onJoin(server, origin, channel)
{
	origin = Util.splituser(origin).toLowerCase();
	channel = channel.toLowerCase();

	write(server, channel, origin);
}

function onMessage(server, origin, channel, message)
{
	origin = Util.splituser(origin).toLowerCase();
	channel = channel.toLowerCase();

	write(server, channel, origin, message);
}

onMe = onMessage;

function onTopic(server, origin, channel)
{
	origin = Util.splituser(origin).toLowerCase();
	channel = channel.toLowerCase();

	write(server, origin, channel)
}

function onLoad()
{
	/*
	 * If the plugin is loaded on-demand, we ask a name list for every
	 * server and every channel of them to update our database.
	 */
	var table = Server.list();

	for (var k in table) {
		var channels = table[k].info().channels;

		for (var i = 0; i < channels.length; ++i) {
			if (channels[i].joined)
				table[k].names(channels[i].name);
		}
	}
}

function onNames(server, channel, list)
{
	for (var i = 0; i < list.length; ++i)
		write(server, channel.toLowerCase(), list[i]);
}
